File: internal/schema/loader.go
package schema

import (
	"fmt"
	"os"
	"strings"

	"gopkg.in/yaml.v3"

	"github.com/aidanlsb/raven/internal/atomicfile"
	"github.com/aidanlsb/raven/internal/paths"
)

// SchemaWarning represents a non-fatal schema issue.
type SchemaWarning struct {
	Message string
}

// LoadResult contains the loaded schema and any warnings.
type LoadResult struct {
	Schema   *Schema
	Warnings []SchemaWarning
}

// Load loads the schema from a vault's schema.yaml file.
// Returns a default schema if the file doesn't exist.
func Load(vaultPath string) (*Schema, error) {
	result, err := LoadWithWarnings(vaultPath)
	if err != nil {
		return nil, err
	}
	return result.Schema, nil
}

// LoadWithWarnings loads the schema and returns any migration warnings.
func LoadWithWarnings(vaultPath string) (*LoadResult, error) {
	schemaPath := paths.SchemaPath(vaultPath)
	result := &LoadResult{Warnings: []SchemaWarning{}}

	if _, err := os.Stat(schemaPath); os.IsNotExist(err) {
		result.Schema = NewSchema()
		return result, nil
	}

	data, err := os.ReadFile(schemaPath)
	if err != nil {
		return nil, fmt.Errorf("failed to read schema file %s: %w", schemaPath, err)
	}

	var schema Schema
	if err := yaml.Unmarshal(data, &schema); err != nil {
		return nil, fmt.Errorf("failed to parse schema file %s: %w", schemaPath, err)
	}

	// Check schema version
	if schema.Version == 0 {
		// No version specified - assume v1 (old format)
		result.Warnings = append(result.Warnings, SchemaWarning{
			Message: "schema.yaml has no version field. Run 'rvn migrate --schema' to upgrade.",
		})
		schema.Version = 1
	} else if schema.Version < CurrentSchemaVersion {
		result.Warnings = append(result.Warnings, SchemaWarning{
			Message: fmt.Sprintf("schema.yaml is version %d, current is %d. Run 'rvn migrate --schema' to upgrade.", schema.Version, CurrentSchemaVersion),
		})
	}

	// Initialize maps if nil
	if schema.Types == nil {
		schema.Types = make(map[string]*TypeDefinition)
	}
	if schema.Core == nil {
		schema.Core = make(map[string]*CoreTypeDefinition)
	}
	if schema.Traits == nil {
		schema.Traits = make(map[string]*TraitDefinition)
	}
	if schema.Templates == nil {
		schema.Templates = make(map[string]*TemplateDefinition)
	}

	// Schema-level templates are file-backed only.
	for templateID, templateDef := range schema.Templates {
		if templateDef == nil {
			return nil, fmt.Errorf("template %q is null; expected an object with at least a file field", templateID)
		}
		spec := strings.TrimSpace(templateDef.File)
		if spec == "" {
			return nil, fmt.Errorf("template %q must define a non-empty file path", templateID)
		}
		if strings.Contains(spec, "\n") || strings.Contains(spec, "\r") {
			return nil, fmt.Errorf("template %q file must be a file path (inline templates are not supported)", templateID)
		}
	}

	// User type template settings must be valid.
	for typeName, typeDef := range schema.Types {
		if typeDef == nil {
			continue
		}
		if IsBuiltinType(typeName) {
			return nil, fmt.Errorf("type %q is a core type; configure it under 'core:' instead of 'types:'", typeName)
		}
		spec := strings.TrimSpace(typeDef.Template)
		if spec != "" {
			if strings.Contains(spec, "\n") || strings.Contains(spec, "\r") {
				return nil, fmt.Errorf("type %q template must be a file path (inline templates are not supported)", typeName)
			}
		}
		if err := validateTemplateBindings(typeName, typeDef.Templates, typeDef.DefaultTemplate, schema.Templates); err != nil {
			return nil, err
		}
	}

	// Core type template settings must be valid.
	for coreName, coreDef := range schema.Core {
		if !IsBuiltinType(coreName) {
			return nil, fmt.Errorf("unknown core type %q under 'core:'", coreName)
		}
		if coreDef == nil {
			return nil, fmt.Errorf("core type %q must be an object", coreName)
		}
		if coreName == "section" {
			if len(coreDef.Templates) > 0 || strings.TrimSpace(coreDef.DefaultTemplate) != "" {
				return nil, fmt.Errorf("core type %q does not support template configuration", coreName)
			}
			continue
		}
		if err := validateTemplateBindings("core."+coreName, coreDef.Templates, coreDef.DefaultTemplate, schema.Templates); err != nil {
			return nil, err
		}
	}

	// Ensure built-in types exist with their fixed definitions.
	// Built-in types are always overwritten to ensure consistency.
	schema.Types["page"] = &TypeDefinition{
		NameField: "title",
		Fields: map[string]*FieldDefinition{
			"title": {Type: FieldTypeString},
		},
	}
	schema.Types["section"] = &TypeDefinition{
		Fields: map[string]*FieldDefinition{
			"title": {Type: FieldTypeString},
			"level": {Type: FieldTypeNumber, Min: floatPtr(1), Max: floatPtr(6)},
		},
	}
	// Built-in 'date' type for daily notes
	dateType := &TypeDefinition{
		Fields: make(map[string]*FieldDefinition),
	}
	if coreDate := schema.Core["date"]; coreDate != nil {
		// Preserve template bindings for built-in date type from core config.
		dateType.Templates = append([]string(nil), coreDate.Templates...)
		dateType.DefaultTemplate = coreDate.DefaultTemplate
	}
	schema.Types["date"] = dateType
	if corePage := schema.Core["page"]; corePage != nil {
		schema.Types["page"].Templates = append([]string(nil), corePage.Templates...)
		schema.Types["page"].DefaultTemplate = corePage.DefaultTemplate
	}

	// Initialize nil field maps for types
	for _, typeDef := range schema.Types {
		if typeDef.Fields == nil {
			typeDef.Fields = make(map[string]*FieldDefinition)
		}
		for _, fieldDef := range typeDef.Fields {
			if fieldDef == nil {
				continue
			}
			fieldDef.Type = normalizeFieldType(fieldDef.Type)
		}
	}
	for _, traitDef := range schema.Traits {
		if traitDef == nil {
			continue
		}
		traitDef.Type = normalizeFieldType(traitDef.Type)
	}

	result.Schema = &schema
	return result, nil
}

func normalizeFieldType(fieldType FieldType) FieldType {
	switch strings.ToLower(string(fieldType)) {
	case "reference":
		return FieldTypeRef
	case "reference[]":
		return FieldTypeRefArray
	case "url":
		return FieldTypeURL
	case "url[]":
		return FieldTypeURLArray
	default:
		return fieldType
	}
}

func validateTemplateBindings(owner string, templateIDs []string, defaultTemplate string, templates map[string]*TemplateDefinition) error {
	seenTemplateIDs := make(map[string]struct{})
	for _, templateID := range templateIDs {
		templateID = strings.TrimSpace(templateID)
		if templateID == "" {
			return fmt.Errorf("%s templates cannot contain empty template IDs", owner)
		}
		if _, seen := seenTemplateIDs[templateID]; seen {
			return fmt.Errorf("%s templates contains duplicate template ID %q", owner, templateID)
		}
		seenTemplateIDs[templateID] = struct{}{}
		if _, ok := templates[templateID]; !ok {
			return fmt.Errorf("%s references unknown template %q", owner, templateID)
		}
	}
	if strings.TrimSpace(defaultTemplate) != "" {
		trimmedDefault := strings.TrimSpace(defaultTemplate)
		if _, ok := seenTemplateIDs[trimmedDefault]; !ok {
			return fmt.Errorf("%s default_template %q is not included in templates", owner, defaultTemplate)
		}
	}
	return nil
}

// CreateDefault creates a default schema.yaml file in the vault.
// Returns true if a new file was created, false if one already existed.
func CreateDefault(vaultPath string) (bool, error) {
	schemaPath := paths.SchemaPath(vaultPath)

	// Skip if file already exists
	if _, err := os.Stat(schemaPath); err == nil {
		return false, nil
	}

	defaultSchema := `# Raven Schema Configuration
# Define your types and traits here.

version: 1  # Schema format version (do not change manually)

# Types: Define what objects ARE (frontmatter 'type:' field)
# Types have fields (structured data in frontmatter)
#
# Built-in types (always available):
#   - page: fallback for files without explicit type
#   - section: auto-created for headings
#   - date: daily notes (files named YYYY-MM-DD.md under directories.daily)
#
# Configure templates for core types under the 'core' block.
# Supported:
#   core.date.templates/default_template
#   core.page.templates/default_template
#   core.section: {}   (placeholder only; no configurable fields)
#
# name_field: When set, 'rvn new <type> <title>' auto-populates this field
# with the title argument. Makes object creation more intuitive.
# description: Optional context for humans/agents (types and fields).

types:
  person:
    description: People and contacts
    default_path: person/
    name_field: name
    fields:
      name:
        type: string
        required: true
      email:
        type: string
        description: Primary contact email

  project:
    default_path: project/
    name_field: name
    fields:
      name:
        type: string
        required: true

  # Example with file-based template:
  # meeting:
  #   default_path: meeting/
  #   templates: [meeting_standard]
  #   default_template: meeting_standard

templates:
  # meeting_standard:
  #   file: templates/meeting.md
  #   description: Standard meeting notes template

core:
  # date:
  #   templates: [daily_default]
  #   default_template: daily_default
  # page:
  #   templates: [note_default]
  #   default_template: note_default
  # section: {}

# Traits: Universal annotations in content (@name or @name(value))
# Traits can be used on any object - just add them to your content.
# Boolean traits: @todo (no value)
# Valued traits: @due(2025-02-01), @priority(high)
traits:
  # Date-related
  due:
    type: date

  # Task marker (boolean trait)
  todo:
    type: boolean

  # Priority
  priority:
    type: enum
    values: [low, medium, high]
    default: medium
`

	if err := atomicfile.WriteFile(schemaPath, []byte(defaultSchema), 0o644); err != nil {
		return false, fmt.Errorf("failed to write schema file: %w", err)
	}

	return true, nil
}

File: src/schema/loader.rs
//! Schema loader - parses schema.yaml

use anyhow::{Context, Result};
use std::path::Path;

use super::Schema;

/// Load schema from a schema.yaml file
pub fn load_schema(vault_path: &Path) -> Result<Schema> {
    let schema_path = vault_path.join("schema.yaml");
    
    if !schema_path.exists() {
        // Return default schema with just the 'page' type
        return Ok(Schema::default());
    }
    
    let contents = std::fs::read_to_string(&schema_path)
        .with_context(|| format!("Failed to read schema file: {}", schema_path.display()))?;
    
    let mut schema: Schema = serde_yaml::from_str(&contents)
        .with_context(|| format!("Failed to parse schema file: {}", schema_path.display()))?;
    
    // Ensure 'page' type exists as fallback
    if !schema.types.contains_key("page") {
        schema.types.insert("page".to_string(), Default::default());
    }
    
    // Ensure 'task' trait has built-in alias if not specified
    if let Some(task_trait) = schema.traits.get_mut("task") {
        if task_trait.cli.is_none() {
            task_trait.cli = Some(super::TraitCliConfig {
                alias: Some("tasks".to_string()),
                default_query: Some("status:todo OR status:in_progress".to_string()),
            });
        }
    }
    
    Ok(schema)
}

/// Create a default schema.yaml file
pub fn create_default_schema(vault_path: &Path) -> Result<()> {
    let schema_path = vault_path.join("schema.yaml");
    
    let default_schema = r#"# Raven Schema Configuration
# Define your types and traits here

types:
  # Example: person type
  # person:
  #   fields:
  #     name:
  #       type: string
  #       required: true
  #     email:
  #       type: string
  #   detect:
  #     path_pattern: "^people/"

  # Example: daily note type
  # daily:
  #   fields:
  #     date:
  #       type: date
  #       derived: from_filename
  #   detect:
  #     path_pattern: "^daily/\\d{4}-\\d{2}-\\d{2}\\.md$"

traits:
  task:
    fields:
      due:
        type: date
      priority:
        type: enum
        values: [low, medium, high]
        default: medium
      status:
        type: enum
        values: [todo, in_progress, done]
        default: todo
    cli:
      alias: tasks
      default_query: "status:todo OR status:in_progress"

  remind:
    fields:
      at:
        type: datetime
        positional: true

  highlight:
    fields:
      color:
        type: enum
        values: [yellow, red, green, blue]
        default: yellow
"#;
    
    std::fs::write(&schema_path, default_schema)
        .with_context(|| format!("Failed to write schema file: {}", schema_path.display()))?;
    
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;
    
    #[test]
    fn test_load_default_schema() {
        let dir = tempdir().unwrap();
        let schema = load_schema(dir.path()).unwrap();
        
        assert!(schema.types.contains_key("page"));
    }
    
    #[test]
    fn test_load_custom_schema() {
        let dir = tempdir().unwrap();
        let schema_content = r#"
types:
  person:
    fields:
      name:
        type: string
        required: true
traits:
  task:
    fields:
      due:
        type: date
"#;
        std::fs::write(dir.path().join("schema.yaml"), schema_content).unwrap();
        
        let schema = load_schema(dir.path()).unwrap();
        
        assert!(schema.types.contains_key("person"));
        assert!(schema.types.contains_key("page")); // Fallback added
        assert!(schema.traits.contains_key("task"));
    }
}
